src/core/tokenizer.c

```diff
@@ -195,3 +197,78 @@ tokenize(filter_pipeline_t *fp, const char *text, size_t text_len __unused)
 	free(content);
 	return NULL;
 }
+
+
+
+/*
+ * tokenize: split the text into tokens separated by a whitespace
+ * and pass the token through the filter pipeline.
+ */
+tokenset_t *
+tokenize2(filter_pipeline_t *fp, const char *text, size_t text_len __unused)
+{
+
+	UBreakIterator *it_token;
+	UChar          utext[text_len + 1]; 
+	UErrorCode     ec = U_ZERO_ERROR;
+	int32_t end;
+	int32_t start;
+	char *content;
+	tokenset_t *tset;
+	char val[10];
+
+	if ((tset = tokenset_create()) == NULL) {
+		return NULL;
+	}
+	if ((content = strdup(text)) == NULL) {
+		goto err;
+	}
+
+	u_uastrcpy(utext, text);
+	it_token = ubrk_open(UBRK_WORD, "en_us", utext, u_strlen(utext), &ec);
+
+	if (__predict_false(U_FAILURE(ec))) {
+		const char *errmsg __unused = u_errorName(ec);
+		app_dbgx("ubrk_open() failed: %s", errmsg);
+		goto err;
+	}
+
+	start = ubrk_first(it_token);
+	for (end = ubrk_next(it_token); end != UBRK_DONE; start = end, end = ubrk_next(it_token)) {
+
+		// XXX: Ignore single chars
+		if (end - start == 1){
+			continue;
+		}
+		strncpy(val, text+start, end-start);
+		val[end-start] = '\0';
```

Comment:
@rmind I suppose this ugly, but didnt know how to do it otherwise

Reply:
@toloco: A few points:
- If you don't have to NIL terminate the string, because `token_create()` takes the length of the string. It makes a copy of string and terminates it in its own buffer.
- What confuses me, though, is: are you operating on UTF-16 or UTF-8? The ICU interface is based on UTF-16, but you take the UTF-16 string ranges and apply them on UTF-8 string. This is wrong for an obvious reason: UTF-16 units by definition might not correspond to a single character. The test happen to succeed because you just use character which map 1:1, but as soon as you'll use a UTF-8 character which is more than 1 byte (e.g. Unicode snowman), it will break.
- Therefore, you have to convert the UTF-16 slice back to UTF-8. It should also use a dynamic buffer, but I suggest make it grow-only i.e. keep re-using it and realloc() if you get a larger string. This way you will minimize the alloc/free.